File: pynanostate/hub-ws.py
from __future__ import print_function
from nanomsg import *

import cherrypy # to host a websocket hub

from ws4py.server.cherrypyserver import WebSocketPlugin, WebSocketTool
from ws4py.websocket import WebSocket
from ws4py.messaging import TextMessage

import argparse

parser = argparse.ArgumentParser(description="A nanostate hub with router between websocket written by Python.")
parser.add_argument('--bind', '-B', dest = "addr", default = "tcp://*:15000", help = "The address to bind, which should be compatible with nanomsg, such as 'tcp://127.0.0.1:15000'. Bind to 'tcp://*:15000' by default.")

parser.add_argument("--ws_host", dest = "ws_host", default = "0.0.0.0", help = "The address to bind websocket server, such as '127.0.0.1'. Bind to '0.0.0.0' (all interfaces) by default.")
parser.add_argument("--ws_port", dest = "ws_port", type=int, default = 15001, help = "The port to bind websocket server, such as '15001'. Bind to '15001' by default.")

args = parser.parse_args()
print(args)
addr = args.addr
ws_host = args.ws_host
ws_port = args.ws_port

class WebSocketHandler(WebSocket):
    def received_message(self, msg):
#        print("WebSocket received: " + repr(msg))
        cherrypy.engine.publish('websocket-broadcast', msg)

        if (msg.is_binary):
            buf = msg.data
            sock.send(buf)
        else:
            sock.send(str(msg))

class Root:
    def __init__(self, addr, port):
        self.addr = addr
        self.port = port

    @cherrypy.expose
    def index(self):
        return "Use " + self.addr + ":" + str(self.port) + " for WebSocket connection."

    @cherrypy.expose
    def ws(self):
        cherrypy.log("Handler created: %s" % repr(cherrypy.request.ws_handler))


# Initialize websocket hub
cherrypy.config.update({
    'server.socket_host': ws_host,
    'server.socket_port': ws_port})
WebSocketPlugin(cherrypy.engine).subscribe()
cherrypy.tools.websocket = WebSocketTool()

cherrypy.tree.mount(Root(ws_host, ws_port), "/", config={
    '/ws': {
        'tools.websocket.on': True,
        'tools.websocket.handler_cls': WebSocketHandler
        }
    })

# Start WebSocket hub
print("Starting WebSocket hub")
cherrypy.engine.start()

# Initialize nanostate hub
sock = Socket(BUS)
sock.bind(args.addr)
sock.send_buffer_size = 1024 * 1024 * 1024
sock.recv_buffer_size = 1024 * 1024 * 1024
sock.set_int_option(SOL_SOCKET, RCVMAXSIZE, -1)
print("Nanostate socket bound to " + addr)

# Start nanostate hub
print("Starting nanostate hub")
while True:
    buf = sock.recv()
#    print("nanostate hub received: " + repr(buf))
    sock.send(buf)
    cherrypy.engine.publish('websocket-broadcast', buf, binary = True)
#    print("WebSocket hub sent: " + repr(buf))

sock.close()
cherrypy.engine.stop()

File: nodenanostate/hub-ws.js
var nano = require("nanomsg");
var argparse = require("argparse")
var WebSocketServer = require("ws").Server

var parser = argparse.ArgumentParser({
  description: "A nanostate hub with route to/from websocket written by Node.js."
});
parser.addArgument(
    ["--bind", "-B"],
    {
      dest: "addr",
      defaultValue: "tcp://127.0.0.1:15000",
      help: "The address to bind, which should be compatible with nanomsg, such as 'tcp://127.0.0.1:15000'"
    });
parser.addArgument(
    ["--ws_host"],
    {
      dest: "ws_host",
      defaultValue: "127.0.0.1",
      help: "The address to bind websocket server, such as 127.0.0.1"
    });
parser.addArgument(
    ["--ws_port"],
    {
      dest: "ws_port",
      type: "int",
      defaultValue: 15001,
      help: "The port to bind websocket server, such as 15001"
    });

var args = parser.parseArgs();
var addr = args.addr;
var ws_host = args.ws_host;
var ws_port = args.ws_port;

// Start socket hub
var sock = nano.socket("bus");
sock.bind(addr);
sock.sndbuf(1024 * 1024 * 1024);
sock.rcvbuf(1024 * 1024 * 1024);
sock.rcvmaxsize(-1);

sock.on("data", function (buf) {
//  console.log("sock received: %s", buf);
//  console.log(sock)
  sock.send(buf);
  wss.broadcast(buf);
});

console.log("Socket bound to " + addr);

// Start WebSocket hub
var wss = new WebSocketServer({
  host: ws_host,
  port: ws_port,
  path: "/ws"
});

wss.broadcast = function broadcast(data) {
  wss.clients.forEach(function each(client) {
    client.send(data, { binary: true });
  });
};

wss.on("connection", function connection(ws) {
  console.log("1 WebSocket client connected.");
  ws.on("message", function(data, flags) {
//    console.log("ws received: %s", data);
    wss.broadcast(data);
    sock.send(data);
  });
});

console.log("WebSocket bound to http://" + ws_host + ":" + ws_port + "/ws");

